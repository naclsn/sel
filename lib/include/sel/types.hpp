#ifndef SEL_TYPE_HPP
#define SEL_TYPE_HPP

/**
 * Parsing and representation of types, as well as a set
 * of helper functions to deal with the type system.
 */

#include <ostream>
#include <istream>
#include <string>
#include <vector>
#include <iterator>


namespace sel {

  enum class Ty {
    UNK,
    NUM,
    STR,
    LST,
    FUN,
  };
  enum TyFlag {
    IS_FIN = 0, // (0, and as such default)
    IS_INF = 1,
    IS_TPL = 2,
  };

  /**
   * Represents a type (yey). Use the [..]Type function
   * to construct/parse.
   */
  struct Type {
    Ty base = Ty::UNK;
    union P {
      std::string* name;
      std::vector<Type*>* box_has; // Ty has;
      Type* box_pair[2]; // Ty pair[2];
    } p = {.name=nullptr};
    uint8_t flags = 0;

    Type() { }
    Type(Ty base, Type::P p, uint8_t flags);
    Type(Type const& ty);
    Type(Type&& ty) noexcept;
    ~Type();

    bool operator==(Type const& other) const;
    bool operator!=(Type const& other) const;

    std::vector<Type*>& has() const { return *p.box_has; }
    Type const& from() const { return *p.box_pair[0]; }
    Type const& to() const { return *p.box_pair[1]; }
    Type applied(Type const& arg) const;
  };

  /**
   * Parse a type from the given stream. The overload to
   * the `>>` operator is also provided as a convenience,
   * but this function allows retreiving a name
   * (which would be specified with the `<name> ::`
   * syntax). If `named` is not nullptr but the parsed
   * type representation does not specify a name, the
   * empty string is written into `named`.
  */
  void parseType(std::istream& in, std::string* named, Type& res);

  std::ostream& operator<<(std::ostream& out, Type const& ty);
  std::istream& operator>>(std::istream& in, Type& res);

} // namespace sel

#endif // SEL_TYPE_HPP
