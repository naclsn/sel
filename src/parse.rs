use std::io;
use std::iter;

#[derive(PartialEq, Debug)]
pub enum Token {
    Unknown(String),
    Word(String),
    Text(Vec<u8>),
    Number(i32),
    Comma,
    OpenBracket,
    CloseBracket,
    OpenBrace,
    CloseBrace,
}

struct Bytes<T: io::Read> {
    inner: io::Bytes<T>,
}

impl<T: io::Read> Iterator for Bytes<T> {
    type Item = u8;

    fn next(&mut self) -> Option<Self::Item> {
        self.inner.next().and_then(|e| e.ok())
    }
}

pub(crate) struct Lexer<T: io::Read> {
    bytes: iter::Peekable<Bytes<T>>,
}

impl<T: io::Read> Lexer<T> {
    fn new(stream: T) -> Self {
        Self {
            bytes: Bytes {
                inner: stream.bytes(),
            }
            .peekable(),
        }
    }
}

impl<T: io::Read> Iterator for Lexer<T> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        Some(match self.bytes.find(|c| !c.is_ascii_whitespace())? {
            b':' => Token::Text({
                iter::from_fn(|| match (self.bytes.next(), self.bytes.peek()) {
                    (Some(b':'), Some(b':')) => {
                        let _ = self.bytes.next();
                        Some(b':')
                    }
                    (Some(b':'), _) => None,
                    (c, _) => c,
                })
                .collect()
            }),

            b',' => Token::Comma,
            b'[' => Token::OpenBracket,
            b']' => Token::CloseBracket,
            b'{' => Token::OpenBrace,
            b'}' => Token::CloseBrace,

            b'#' => self.bytes.find(|&c| b'\n' == c).and_then(|_| self.next())?,

            c if c.is_ascii_lowercase() => Token::Word(
                String::from_utf8(
                    iter::once(c)
                        .chain(iter::from_fn(|| self.bytes.next_if(u8::is_ascii_lowercase)))
                        .collect(),
                )
                .unwrap(),
            ),

            c if c.is_ascii_digit() => Token::Number({
                let mut r = 0i32;
                let (shift, digits) = match (c, self.bytes.peek()) {
                    (b'0', Some(b'B' | b'b')) => {
                        self.bytes.next();
                        (1, b"01".as_slice())
                    }
                    (b'0', Some(b'O' | b'o')) => {
                        self.bytes.next();
                        (3, b"01234567".as_slice())
                    }
                    (b'0', Some(b'X' | b'x')) => {
                        self.bytes.next();
                        (4, b"0123456789abcdef".as_slice())
                    }
                    _ => {
                        r = c as i32 - 48;
                        (0, b"0123456789".as_slice())
                    }
                };
                while let Some(k) = self
                    .bytes
                    .peek()
                    .and_then(|&c| digits.iter().position(|&k| k == c | 32))
                {
                    self.bytes.next();
                    r = if 0 == shift {
                        r * 10 + k as i32
                    } else {
                        r << shift | k as i32
                    }
                }
                r
            }),

            c => Token::Unknown(
                String::from_utf8_lossy(
                    &iter::once(c)
                        .chain(self.bytes.by_ref().take_while(|c| !c.is_ascii_whitespace()))
                        .collect::<Vec<_>>(),
                )
                .into_owned(),
            ),
        })
    }
}

#[derive(PartialEq, Debug)]
pub enum Tree {
    Apply(Box<Tree>, Vec<Tree>),
    Atom(Token),
    Chain(Vec<Tree>),
    List(Vec<Tree>),
}

#[derive(PartialEq, Debug)]
pub enum Error {
    Unexpected(Token),
    UnexpectedEOF,
}

impl Tree {
    fn one_from_peekable(
        peekable: &mut iter::Peekable<Lexer<impl io::Read>>,
    ) -> Result<Tree, Error> {
        let one = match peekable.next() {
            Some(token @ (Token::Word(_) | Token::Text(_) | Token::Number(_))) => Tree::Atom(token),

            Some(Token::OpenBracket) => Tree::Chain(
                iter::once(Tree::one_from_peekable(peekable))
                    .chain(iter::from_fn(|| match peekable.next() {
                        Some(Token::Comma) => Some(Tree::one_from_peekable(peekable)),
                        Some(Token::CloseBracket) => None,
                        Some(token) => Some(Err(Error::Unexpected(token))),
                        None => Some(Err(Error::UnexpectedEOF)),
                    }))
                    .collect::<Result<_, _>>()?,
            ),

            Some(Token::OpenBrace) => Tree::List(
                iter::once(Tree::one_from_peekable(peekable))
                    .chain(iter::from_fn(|| match peekable.next() {
                        Some(Token::Comma) => Some(Tree::one_from_peekable(peekable)),
                        Some(Token::CloseBrace) => None,
                        Some(token) => Some(Err(Error::Unexpected(token))),
                        None => Some(Err(Error::UnexpectedEOF)),
                    }))
                    .collect::<Result<_, _>>()?,
            ),

            Some(token) => return Err(Error::Unexpected(token)),
            None => return Err(Error::UnexpectedEOF),
        };

        let args: Vec<_> = iter::from_fn(|| match peekable.peek() {
            Some(Token::Comma | Token::CloseBracket | Token::CloseBrace) | None => None,
            _ => Some(Tree::one_from_peekable(peekable)),
        })
        .collect::<Result<_, _>>()?;
        Ok(if args.is_empty() {
            one
        } else {
            Tree::Apply(Box::new(one), args)
        })
    }

    fn from_peekable(peekable: &mut iter::Peekable<Lexer<impl io::Read>>) -> Result<Tree, Error> {
        let mut r = vec![Tree::one_from_peekable(peekable)?];
        while let Some(Token::Comma) = peekable.next() {
            r.push(Tree::one_from_peekable(peekable)?);
        }
        Ok(Tree::Chain(r))
    }

    pub fn from_lexer(lexer: Lexer<impl io::Read>) -> Result<Tree, Error> {
        Tree::from_peekable(&mut lexer.peekable())
    }

    pub fn from_stream(stream: impl io::Read) -> Result<Tree, Error> {
        Tree::from_lexer(Lexer::new(stream))
    }

    pub fn from_str(text: &str) -> Result<Tree, Error> {
        Tree::from_stream(text.as_bytes())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn lexing() {
        assert_eq!(
            &[Token::Word("coucou".into()),][..],
            Lexer::new("coucou".as_bytes()).collect::<Vec<Token>>()
        );

        assert_eq!(
            &[
                Token::OpenBracket,
                Token::Word("a".to_string()),
                Token::Number(1),
                Token::Word("b".to_string()),
                Token::Comma,
                Token::OpenBrace,
                Token::Word("w".to_string()),
                Token::Comma,
                Token::Word("x".to_string()),
                Token::Comma,
                Token::Word("y".to_string()),
                Token::Comma,
                Token::Word("z".to_string()),
                Token::CloseBrace,
                Token::CloseBracket,
            ][..],
            Lexer::new("[a 1 b, {w, x, y, z}]".as_bytes()).collect::<Vec<Token>>()
        );

        assert_eq!(
            &[
                Token::Number(42),
                Token::Text(vec![42]),
                Token::Number(42),
                Token::Number(42),
                Token::Number(42),
            ][..],
            Lexer::new("42 :*: 0x2a 0b101010 0o52".as_bytes()).collect::<Vec<Token>>()
        );
    }

    #[test]
    fn parsing() {
        assert_eq!(
            Ok(Tree::Chain(vec![Tree::Atom(Token::Word("coucou".to_string()))])),
            Tree::from_str("coucou")
        );

        assert_eq!(
            Ok(Tree::Chain(vec![
                Tree::Apply(
                    Box::new(Tree::Atom(Token::Word("a".to_string()))),
                    vec![Tree::Apply(
                        Box::new(Tree::Atom(Token::Number(1))),
                        vec![Tree::Atom(Token::Word("b".to_string()))],
                    )],
                ),
                Tree::List(vec![
                    Tree::Atom(Token::Word("w".to_string())),
                    Tree::Atom(Token::Word("x".to_string())),
                    Tree::Atom(Token::Word("y".to_string())),
                    Tree::Atom(Token::Word("z".to_string())),
                ])
            ])),
            Tree::from_str("a 1 b, {w, x, y, z}")
        );

        // todo
        //assert_eq!(Err...)
    }
}
