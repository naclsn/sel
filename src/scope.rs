use std::collections::HashMap;
use std::fs::File;
use std::io::{Read, Result as IoResult};
use std::mem;
use std::ops::Range;
use std::path::{Path, PathBuf};
use std::ptr;

use crate::builtin;
use crate::parse::Tree;
use crate::types::{TypeList, TypeRef};

pub struct Global {
    pub registry: SourceRegistry,
    pub types: TypeList,
    pub scope: Scope,
}

#[derive(Debug)]
pub enum ScopeItem {
    Builtin(fn(&mut TypeList) -> TypeRef, &'static str),
    Defined(Tree, String),
    Binding(Location, TypeRef),
}

#[derive(Debug)]
pub struct Scope {
    parent: *const Scope,
    names: HashMap<String, ScopeItem>,
}

// source registry {{{
pub type SourceRef = usize;

#[derive(PartialEq, Debug, Clone)]
pub struct Location(pub SourceRef, pub Range<usize>);

#[derive(Default)]
pub struct SourceRegistry(Vec<(PathBuf, Option<Vec<u8>>)>);

impl SourceRegistry {
    /// Note: the given path is taken as is (ie not canonicalize) and not checked for duplication
    pub fn add_bytes(&mut self, name: impl AsRef<Path>, bytes: Vec<u8>) -> SourceRef {
        self.0.push((name.as_ref().to_owned(), Some(bytes)));
        self.0.len() - 1
    }

    pub fn add(&mut self, name: impl AsRef<Path>) -> IoResult<SourceRef> {
        let name = name.as_ref().canonicalize()?;
        if let Some(found) = self.0.iter().position(|c| c.0 == name) {
            return Ok(found);
        }
        let mut bytes = Vec::new();
        File::open(&name)?.read_to_end(&mut bytes)?;
        Ok(self.add_bytes(name, bytes))
    }

    pub fn get_path(&self, at: SourceRef) -> &Path {
        &self.0[at].0
    }

    pub fn get_bytes(&self, at: SourceRef) -> &[u8] {
        self.0[at].1.as_ref().unwrap()
    }

    pub fn take_bytes(&mut self, at: SourceRef) -> Vec<u8> {
        self.0[at].1.take().unwrap()
    }

    pub fn put_back_bytes(&mut self, at: SourceRef, bytes: Vec<u8>) {
        self.0[at].1 = Some(bytes);
    }
}
// }}}

impl Global {
    pub fn with_builtin() -> Global {
        Global {
            registry: SourceRegistry::default(),
            types: TypeList::default(),
            scope: builtin::scope(),
        }
    }
}

// scope {{{
impl ScopeItem {
    pub fn make_type(&self, types: &mut TypeList) -> TypeRef {
        match self {
            ScopeItem::Builtin(mkty, _) => mkty(types),
            ScopeItem::Defined(val, _) => types.duplicate(val.ty, &mut HashMap::new()),
            ScopeItem::Binding(_, ty) => types.duplicate(*ty, &mut HashMap::new()),
        }
    }

    pub fn get_loc(&self) -> Option<&Location> {
        match self {
            ScopeItem::Builtin(_, _) => None,
            ScopeItem::Defined(val, _) => Some(&val.loc),
            ScopeItem::Binding(loc, _) => Some(loc),
        }
    }

    pub fn get_desc(&self) -> Option<&str> {
        match self {
            ScopeItem::Builtin(_, desc) => Some(desc),
            ScopeItem::Defined(_, desc) => Some(desc),
            ScopeItem::Binding(_, _) => None,
        }
    }
}

// YYY: weird pattern, surely there is better :/
pub(crate) struct MustRestore(Scope);

impl Default for Scope {
    fn default() -> Scope {
        Scope {
            parent: ptr::null(),
            names: HashMap::default(),
        }
    }
}

impl Scope {
    pub(crate) fn new(parent: &Scope) -> Scope {
        Scope {
            parent: parent as *const Scope,
            names: HashMap::new(),
        }
    }

    /// returns the parent scope (previously `self`)
    /// which is expected to be restored with `restore_from_parent`
    pub(crate) fn make_into_child(&mut self) -> MustRestore {
        let parent = mem::take(self);
        self.parent = &parent as *const Scope;
        MustRestore(parent)
    }
    /// see `make_into_child`
    pub(crate) fn restore_from_parent(&mut self, parent: MustRestore) {
        *self = parent.0;
    }

    /// parent-most englobing scope
    pub fn global(&self) -> &Scope {
        unsafe { self.parent.as_ref() }.unwrap_or(self)
    }

    /// does not insert if the name already existed, returns a reference to the previous value
    pub(crate) fn declare(&mut self, name: String, value: ScopeItem) -> Option<&ScopeItem> {
        let mut already = false;
        let r = self
            .names
            .entry(name)
            .and_modify(|_| already = true)
            .or_insert_with(|| value);
        if already {
            Some(r)
        } else {
            None
        }
    }

    pub fn lookup(&self, name: &str) -> Option<&ScopeItem> {
        self.names
            .get(name)
            .or_else(|| unsafe { self.parent.as_ref() }?.lookup(name))
    }

    pub fn iter(&self) -> <&HashMap<String, ScopeItem> as IntoIterator>::IntoIter {
        self.names.iter()
    }

    pub fn iter_rec(&self) -> ScopeRecIter {
        ScopeRecIter {
            entries: self.names.iter(),
            parent: unsafe { self.parent.as_ref() },
        }
    }
}

// iters {{{
pub struct ScopeRecIter<'a> {
    entries: <&'a HashMap<String, ScopeItem> as IntoIterator>::IntoIter,
    parent: Option<&'a Scope>,
}

impl<'a> Iterator for ScopeRecIter<'a> {
    type Item = (&'a String, &'a ScopeItem);

    fn next(&mut self) -> Option<Self::Item> {
        self.entries.next().or_else(|| {
            *self = self.parent?.iter_rec();
            self.next()
        })
    }
}

impl IntoIterator for Scope {
    type IntoIter = <HashMap<String, ScopeItem> as IntoIterator>::IntoIter;
    type Item = <Self::IntoIter as Iterator>::Item;

    fn into_iter(self) -> Self::IntoIter {
        self.names.into_iter()
    }
}

impl<'a> IntoIterator for &'a Scope {
    type IntoIter = <&'a HashMap<String, ScopeItem> as IntoIterator>::IntoIter;
    type Item = <Self::IntoIter as Iterator>::Item;

    fn into_iter(self) -> Self::IntoIter {
        self.names.iter()
    }
}
// }}}
// }}}
