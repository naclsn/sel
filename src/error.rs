use std::fmt::{Display, Formatter, Result as FmtResult};

use crate::parse::{Applicable, TokenKind};
use crate::scope::{Location, SourceRegistry};
use crate::types::FrozenType;

// error types {{{
#[derive(PartialEq, Debug)]
pub enum ErrorContext {
    Unmatched {
        open_token: TokenKind,
    },
    CompleteType {
        complete_type: FrozenType,
    },
    TypeListInferredItemType {
        list_item_type: FrozenType,
    },
    AsNthArgToNowTyped {
        nth_arg: usize,
        func: Applicable,
        type_with_curr_args: FrozenType,
    },
    ChainedFromAsNthArgToNowTyped {
        comma_loc: Location,
        nth_arg: usize,
        func: Applicable,
        type_with_curr_args: FrozenType,
    },
    ChainedFromToNotFunc {
        comma_loc: Location,
    },
    AutoCoercedVia {
        func_name: String,
        func_type: FrozenType,
    },
    DeclaredHereWithType {
        with_type: FrozenType,
    },
    LetFallbackTypeMismatch {
        result_type: FrozenType,
        fallback_type: FrozenType,
    },
}

#[derive(PartialEq, Debug)]
pub enum ErrorKind {
    ContextCaused {
        error: Box<Error>,
        because: ErrorContext,
    },
    Unexpected {
        token: TokenKind,
        expected: &'static str,
    },
    UnexpectedDefInScript,
    UnknownName {
        name: String,
        expected_type: FrozenType,
    },
    NotFunc {
        actual_type: FrozenType,
    },
    TooManyArgs {
        nth_arg: usize,
        func: Applicable,
    },
    ExpectedButGot {
        expected: FrozenType,
        actual: FrozenType,
    },
    InfWhereFinExpected,
    NameAlreadyDeclared {
        name: String,
    },
    CouldNotReadFile {
        error: String,
    },
}

#[derive(PartialEq, Debug)]
pub struct Error(pub Location, pub ErrorKind);

#[derive(Default, PartialEq, Debug)]
pub struct ErrorList(pub(crate) Vec<Error>);

impl IntoIterator for ErrorList {
    type Item = Error;
    type IntoIter = <Vec<Error> as IntoIterator>::IntoIter;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}

impl ErrorList {
    pub(crate) fn push(&mut self, err: Error) {
        self.0.push(err);
    }

    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}
// }}}

pub struct Report<'a> {
    registry: &'a SourceRegistry,
    title: String,
    messages: Vec<(Location, String)>,
}

// generate report {{{
impl Error {
    fn ctx_messages(loc: Location, because: &ErrorContext, report: &mut Report) {
        use ErrorContext::*;
        use TokenKind::*;

        let msgs: &[_] = match because {
            Unmatched { open_token } => &[(
                loc,
                format!(
                    "open {} here",
                    match open_token {
                        OpenBracket => '[',
                        OpenBrace => '{',
                        Unknown(_) | Word(_) | Bytes(_) | Number(_) | Comma | CloseBracket
                        | CloseBrace | Equal | Def | Let | Use | Semicolon | End => unreachable!(),
                    }
                ),
            )],
            CompleteType { complete_type } => &[(loc, format!("complete type: {complete_type}"))],
            TypeListInferredItemType { list_item_type } => &[(
                loc,
                format!("list type was inferred to be [{list_item_type}]"),
            )],
            AsNthArgToNowTyped {
                nth_arg,
                func,
                type_with_curr_args,
            } => &[(
                loc,
                format!(
                    "see parameter in {type_with_curr_args} (overall {nth_arg}{} argument to {})",
                    match nth_arg {
                        1 => "st",
                        2 => "nd",
                        3 => "rd",
                        _ => "th",
                    },
                    match func {
                        Applicable::Name(name) => name,
                        Applicable::Bind(_, _, _) => "let binding",
                    },
                ),
            )],
            ChainedFromAsNthArgToNowTyped {
                comma_loc,
                nth_arg,
                func,
                type_with_curr_args,
            } => &[
                (
                    loc,
                    format!(
                        "see parameter in {type_with_curr_args} (overall {nth_arg}{} argument to {})",
                        match nth_arg {
                            1 => "st",
                            2 => "nd",
                            3 => "rd",
                            _ => "th",
                        },
                        match func {
                            Applicable::Name(name) => name,
                            Applicable::Bind(_, _, _) => "let binding",
                        },
                    ),
                ),
                (comma_loc.clone(), "chained through here".into()),
            ],
            ChainedFromToNotFunc { comma_loc } => &[
                (loc, "Not a function".into()),
                (comma_loc.clone(), "chained through here".into()),
            ],
            AutoCoercedVia {
                func_name,
                func_type,
            } => &[(
                loc,
                format!("coerced via '{func_name}' (which has type {func_type})"),
            )],
            DeclaredHereWithType { with_type } => &[(loc, format!("here with type {with_type}"))],
            LetFallbackTypeMismatch {
                result_type,
                fallback_type,
            } => &[(
                loc,
                format!("fallback of type {fallback_type} doesn't match result type {result_type}"),
            )],
        };
        report.messages.extend_from_slice(msgs);
    }

    pub fn report<'a>(&self, registry: &'a SourceRegistry) -> Report<'a> {
        use ErrorKind::*;
        use TokenKind::*;

        let loc = self.0.clone();

        match &self.1 {
            ContextCaused { error, because } => {
                let mut r = error.report(registry);
                Error::ctx_messages(loc, because, &mut r);
                r
            }
            Unexpected { token, expected } => Report {
                registry,
                title: "Unexpected token".into(),
                messages: vec![(
                    loc,
                    format!(
                        "Unexpected '{}', expected {expected}",
                        match token {
                            Unknown(t) => t,
                            Word(v) => v,
                            Bytes(_) => "(bytes)",
                            Number(_) => "(number)",
                            Comma => ",",
                            OpenBracket => "[",
                            CloseBracket => "]",
                            OpenBrace => "{",
                            CloseBrace => "}",
                            Equal => "=",
                            Def => "def",
                            Let => "let",
                            Use => "use",
                            Semicolon => ";",
                            End => "end of script",
                        }
                    ),
                )],
            },
            UnexpectedDefInScript => Report {
                registry,
                title: "Unexpected definition".into(),
                messages: vec![(
                    loc,
                    "these should only appear before the script, at the beginning of the file"
                        .into(),
                )],
            },
            UnknownName {
                name,
                expected_type,
            } => Report {
                registry,
                title: "Unknown name".into(),
                messages: vec![(
                    loc,
                    format!("Unknown name '{name}', should be of type {expected_type}"),
                )],
            },
            NotFunc { actual_type } => Report {
                registry,
                title: "Not a function".into(),
                messages: vec![(
                    loc,
                    format!("Expected a function type, but got {actual_type}"),
                )],
            },
            TooManyArgs { nth_arg, func } => Report {
                registry,
                title: "Too many argument".into(),
                messages: vec![(
                    loc,
                    format!(
                        "Too many argument to {}, expected only {}",
                        match func {
                            Applicable::Name(name) => name,
                            Applicable::Bind(_, _, _) => "let binding",
                        },
                        nth_arg - 1
                    ),
                )],
            },
            ExpectedButGot { expected, actual } => Report {
                registry,
                title: "Type mismatch".into(),
                messages: vec![(loc, format!("Expected type {expected}, but got {actual}"))],
            },
            InfWhereFinExpected => Report {
                registry,
                title: "Wrong boundedness".into(),
                messages: vec![(loc, "Expected finite type, but got infinite type".into())],
            },
            NameAlreadyDeclared { name } => Report {
                registry,
                title: "Name used twice".into(),
                messages: vec![(loc, format!("Name {name} was already declared"))],
            },
            CouldNotReadFile { error } => Report {
                registry,
                title: "Could not read file".into(),
                messages: vec![(loc, format!("<< {error} >>"))],
            },
        }
    }
}
// }}}

// display report {{{
impl Display for Report<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        let (Location(file, ref range), _) = self.messages[0];
        let path = self.registry.get_path(file);
        let bytes = self.registry.get_bytes(file);
        let line = 1 + bytes[range.start..].iter().filter(|c| b'\n' == **c).count();
        writeln!(f, "{}:{line}: {}", path.display(), self.title)?;

        for (_, msg) in &self.messages {
            writeln!(f, " -> {}", msg)?;
        }
        Ok(())
    }
}
// }}}
