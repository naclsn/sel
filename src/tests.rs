use insta::{assert_debug_snapshot, assert_snapshot};

use crate::error::ErrorList;
use crate::parse::{process, Lexer, TokenKind, Tree};
use crate::scope::Global;
use crate::types::FrozenType;

#[test]
fn lexing() {
    fn t(script: &[u8]) -> String {
        let mut end = true;
        Lexer::new(0, script.iter().copied())
            .take_while(|t| {
                if !end {
                    false
                } else {
                    end = TokenKind::End != t.1;
                    true
                }
            })
            .fold(String::new(), |mut acc, cur| {
                acc += &format!("{cur:?}\n");
                acc
            })
    }

    assert_snapshot!(t(b""));
    assert_snapshot!(t(b"coucou"));
    assert_snapshot!(t(b"[a 1 b, {w, x, y, z}, 0.5]"));
    assert_snapshot!(t(b":hay: :hey:: not hay: :: :::: fin"));
    assert_snapshot!(t(b"42 :*: 0x2a 0b101010 0o52"));
    assert_snapshot!(t(b"oh-no; don't let use be the def of cat!"));
}

#[test]
fn parsing() {
    fn t(script: &[u8]) -> Result<(FrozenType, Tree), ErrorList> {
        let mut global = Global::with_builtin();
        let source = global.registry.add_bytes("<test>", script.into());
        let result = process(source, &mut global);
        result
            .tree
            .map(|t| (global.types.frozen(t.ty), t))
            .ok_or(result.errors)
    }

    assert_debug_snapshot!(t(b""));
    assert_debug_snapshot!(t(b"-"));
    assert_debug_snapshot!(t(b"-, split:-:, map[tonum, add1, tostr], join:+:"));
    assert_debug_snapshot!(t(b"tonum, add234121, tostr, ln"));
    assert_debug_snapshot!(t(b"[tonum, add234121, tostr] :13242:"));
    assert_debug_snapshot!(t(b"{repeat 1, {}}"));
    assert_debug_snapshot!(t(b"add 1, tostr"));
    assert_debug_snapshot!(t(b"zipwith add {1}"));
    assert_debug_snapshot!(t(b"{repeat 1, {}}"));
    assert_debug_snapshot!(t(b"{{}, repeat 1}"));
    assert_debug_snapshot!(t(b"{{42}, repeat 1}"));
    assert_debug_snapshot!(t(b"{1, 2, 3, :soleil:}, ln"));
    assert_debug_snapshot!(t(b"add 1 :2:"));
    assert_debug_snapshot!(t(b":42:, add 1"));
    assert_debug_snapshot!(t(b"add 1, tostr"));
    assert_debug_snapshot!(t(b"add 1, tonum")); // TODO: incorrect if coercion should occure
    assert_debug_snapshot!(t(b"prout 1, caca"));
    assert_debug_snapshot!(t(b"_, split_, map_, join_"));
    assert_debug_snapshot!(t(b"{const, _, add, map}"));
    assert_debug_snapshot!(t(b"2 -"));
    assert_debug_snapshot!(t(b"- 2"));
    assert_debug_snapshot!(t(b"add 1 2 3"));
    assert_debug_snapshot!(t(b"3, add 1 2"));
    assert_debug_snapshot!(t(b":1:, tostr, _"));
    assert_debug_snapshot!(t(b"iterate [_ 1, add 1] 1")); // _ :: Num -> Num -> Num
    assert_debug_snapshot!(t(b"iterate [div 1, _ 1] 1")); // TODO: _ :: Num -> Num -> Num
    assert_debug_snapshot!(t(b"iterate [div 1, add 1] 1, _")); // _ :: [Num]+ -> returnof(typeof(_))
    assert_debug_snapshot!(t(b"let {a, b} [add a b]"));
    assert_debug_snapshot!(t(b"{1, 2, 3}, let {h,, t} h"));
    assert_debug_snapshot!(t(b"{1, 2, 3}, let {h,, t} t"));
    assert_debug_snapshot!(t(b"repeat 1, let {h,, t} t"));
    assert_debug_snapshot!(t(b"let a let b a"));
}
