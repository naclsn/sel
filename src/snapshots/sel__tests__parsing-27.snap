---
source: src/tests.rs
expression: "t(b\"iterate [div 1, add 1] 1, _\")"
---
Err(
    ErrorList(
        [
            Error(
                Location(
                    0,
                    26..27,
                ),
                UnknownName {
                    name: "_",
                    expected_type: Func(
                        List(
                            false,
                            Number,
                        ),
                        Named(
                            "returnof(typeof(_))",
                        ),
                    ),
                },
            ),
        ],
    ),
)
